commit f8e4a6971e9bc94cee1e988fa9943475112917e9
Author: Joe Thornber <ejt@redhat.com>
Date:   Wed Dec 11 15:24:39 2024 +0000

    [dm-unit] dm-array fixes patch 3

diff --git a/drivers/md/persistent-data/dm-array.c b/drivers/md/persistent-data/dm-array.c
index 0850dfdffc8c..8f8792e55806 100644
--- a/drivers/md/persistent-data/dm-array.c
+++ b/drivers/md/persistent-data/dm-array.c
@@ -1003,6 +1003,7 @@ int dm_array_cursor_skip(struct dm_array_cursor *c, uint32_t count)
 		}
 
 		count -= remaining;
+		c->index += (remaining - 1);
 		r = dm_array_cursor_next(c);
 
 	} while (!r);
